influx/writer.go

```diff
@@ -100,38 +113,87 @@ func writeInBatches(log ttnlog.Interface, w BatchPointWriter, bpConf influxdb.Ba
 	}
 }
 
-type batchingWriter struct {
+// BatchingWriter is a PointWriter, which writes points in batches.
+// BatchingWriter scales automatically once it notices a delay of scalingInterval to write a batch of points and downscales if no points are supplied to an instance for a duration of scalingInterval.
+// BatchingWriter spawns an instance for each unique BatchPointsConfig specified and up to limit() additional instances on top of that.
+// BatchingWriter does not limit the amount of instances if limit is nil.
+// Maximum number of instances spawned is equal to amount of unique BatchPointsConfig passed plus value, specified by WithInstanceLimit option.
+// By default, BatchingWriter does not limit amount of instances.
+// Each instance is spawned in a separate goroutine.
+type BatchingWriter struct {
 	log             ttnlog.Interface
+	writer          BatchPointsWriter
 	scalingInterval time.Duration
 
-	writer     BatchPointWriter
-	mutex      sync.RWMutex
-	pointChans map[influxdb.BatchPointsConfig]chan *batchPoint
+	activeMutex sync.RWMutex
+	active      uint
+	limitMutex  sync.RWMutex
+	limit       uint
+
+	pointChanMutex sync.RWMutex
+	pointChans     map[influxdb.BatchPointsConfig]chan *batchPoint
+}
+
+// BatchingWriterOption is passed to the constructor of BatchingWriter to configure it accordingly
+type BatchingWriterOption func(w *BatchingWriter)
+
+// WithInstanceLimit sets a limit on amount of additional instances spawned by BatchingWriter
+func WithInstanceLimit(v uint) BatchingWriterOption {
+	return func(w *BatchingWriter) {
+		w.limit = v
+	}
+}
+
+// WithInstanceLimit sets a limit on amount of additional instances spawned by BatchingWriter
+func WithScalingInterval(v time.Duration) BatchingWriterOption {
+	return func(w *BatchingWriter) {
+		w.scalingInterval = v
+	}
 }
 
-// NewBatchingWriter creates new PointWriter, which writes points in batches and scales automatically according to scalingInterval.
-func NewBatchingWriter(log ttnlog.Interface, w BatchPointWriter, scalingInterval time.Duration) PointWriter {
-	return &batchingWriter{
+// NewBatchingWriter creates new BatchingWriter. If WithScalingInterval is not specified, DefaultScalingInterval value is used. If WithInstanceLimit is not specified, DefaultInstanceLimit is used.
+func NewBatchingWriter(log ttnlog.Interface, w BatchPointsWriter, opts ...BatchingWriterOption) *BatchingWriter {
+	bw := &BatchingWriter{
 		log:             log,
 		writer:          w,
-		scalingInterval: scalingInterval,
+		scalingInterval: DefaultScalingInterval,
+		limit:           DefaultInstanceLimit,
```

Comment:
Maybe we can just use `GOMAXPROCS` as a default limit. This will cover cases where you run the database on the same machine as the writer. When the db is on a different machine, you probably won't use the defaults anyway